assets/src/pages/room/hooks/useAutomaticEncodingSwitching.tsx

```diff
@@ -0,0 +1,79 @@
+import { MembraneWebRTC, TrackEncoding } from "@jellyfish-dev/membrane-webrtc-js";
+import { useCallback, useEffect, useState } from "react";
+import { useResizeDetector } from "react-resize-detector";
+import { useStoreFirstNonNullValue } from "./useStoreFirstNonNullValue";
+import {
+  MAX_TILE_HEIGHT_FOR_LOW_ENCODING,
+  MAX_TILE_HEIGHT_FOR_MEDIUM_ENCODING,
+  VIDEO_TILE_RESIZE_DETECTOR_DEBOUNCE_VALUE,
+} from "../consts";
+import { useSimulcastRemoteEncoding } from "./useSimulcastRemoteEncoding";
+
+const getHighestAllowedEncoding = (height: number | null): TrackEncoding | null => {
+  if (height === null) return null;
+  if (height < MAX_TILE_HEIGHT_FOR_LOW_ENCODING) return "l";
+  if (height < MAX_TILE_HEIGHT_FOR_MEDIUM_ENCODING) return "m";
+  return "h";
+};
+
+const useCalculateSmartEncoding = (forceEncoding: TrackEncoding | null) => {
+  const { height, ref } = useResizeDetector({
+    refreshMode: "debounce",
+    refreshRate: VIDEO_TILE_RESIZE_DETECTOR_DEBOUNCE_VALUE,
+    handleWidth: false,
+  });
+  const tileSizeEncoding = getHighestAllowedEncoding(height || null);
+  const smartEncoding: TrackEncoding | null = forceEncoding ?? tileSizeEncoding;
+
+  return { ref, tileSizeEncoding, smartEncoding };
+};
+
+export const useAutomaticEncodingSwitching = (
+  currentTrackEncoding: TrackEncoding | null,
+  peerId: string | null,
+  trackId: string | null,
+  disableAutomaticLayerSwitching: boolean,
+  forceEncoding: TrackEncoding | null,
+  webrtc: MembraneWebRTC | null
+) => {
+  const [smartEncodingStatus, setSmartEncodingStatus] = useState(true);
+  const onInitEncodingQuality = useStoreFirstNonNullValue(currentTrackEncoding || null);
+
+  const { ref, tileSizeEncoding, smartEncoding } = useCalculateSmartEncoding(forceEncoding);
+
+  const { setTargetEncoding, targetEncoding } = useSimulcastRemoteEncoding(peerId, trackId, webrtc);
+
+  useEffect(() => {
+    if (disableAutomaticLayerSwitching) return;
+    if (!smartEncodingStatus) return;
+    if (!onInitEncodingQuality) return;
+    if (!smartEncoding) return;
```

Comment:
Imo it's more readable when it is in one `if` with a statement containing logical alternatives